crates/noirc_evaluator/src/ssa/code_gen.rs

```diff
@@ -47,6 +46,10 @@ impl Value {
             Value::Struct(_) => panic!("Tried to unwrap a struct into a single value"),
         }
     }
+
+    pub fn dummy() -> Value {
+        Value::Single(NodeId::dummy())
```

Comment:
and it is better if this is handled by the frontend.

Reply:
Clearly the return statement is a frontend thing! If I write a function that returns a value with no return at the end the frontend is complaining. Furthermore when we will support real return statements we will not need to add a new way to handle return statements in the IR.

Reply:
If you write a function with no `return` the program will work as intended because `return` isn't in the language. The frontend never complains about missing returns, only incorrect return types. Inserting the Return operation into ssa is a frontend thing - it is done in the code_gen pass of the frontend. Once/if we support early return operations, the ssa pass will need to change anyway to support multiple returns from a function. The best way to do this is probably to create an end block and jump there with a phi of all return values. Existing IRs like llvm allow multiple return instructions in separate blocks and do any conversions/analyses internally but I'd be open to doing it in the code_gen pass as well.

Reply:
I agree with you, but my point is since the frontend knows about return values (because it knows their type), let's have it put a return statement in the AST.
Codegen will parse this return statement, and so it will when the language will have return statements. Obviously, we will have to modify the way we handle them, but at least not the way it parses them.
I think it is much cleaner if the frontend takes care of return statements, whether they are or not in the language, and abstract this away from codegen.

Reply:
Once we have early returns in the syntax there will still need to be special handling added to the ssa IR. This must be handled in if statements and anywhere else a Jmp-like (or existing Ret) instruction is created to avoid adding a jump instruction after a return instruction was already added. From the frontend's perspective there is nothing clear to do as it is the SSA pass' job to insert Jmp instructions (thus it must be the one to _not_ insert them when needed). Thus even if the frontend handles adding a return statement manually it does not help the ssa pass much.

I'll also refer to the llvm codegen pass of my compiler here for functions: https://github.com/jfecher/ante/blob/master/src/llvm/mod.rs#L502 which does not need any special casing to add returns (aside from the case mentioned above checking the block isn't already terminated).